kotlin-inject-compiler/ksp/src/main/kotlin/me/tatarka/inject/compiler/ksp/InjectProcessor.kt

```diff
@@ -27,63 +28,69 @@ class InjectProcessor(
 
-        val kmpComponentCreateSymbols = previousDeferredFunctions + resolver.getSymbolsWithFunctionAnnotation(
-            packageName = KMP_COMPONENT_CREATE.packageName,
-            simpleName = KMP_COMPONENT_CREATE.simpleName
-        )
-        deferredFunctions = kmpComponentCreateSymbols.filterNot { element ->
+        val kmpComponentCreateSymbols = resolver.getSymbolsWithAnnotation(KMP_COMPONENT_CREATE.canonicalName)
+            .mapNotNull { it as? KSFunctionDeclaration }
+        val deferredFunctions = kmpComponentCreateSymbols.filterNot { element ->
             processKmpComponentCreate(element, provider, kmpComponentCreateFunctionsByComponentType)
-        }
+        }.toList()
+        deferredFunctionNames = deferredFunctions.mapNotNull { it.qualifiedName }
 
         return deferredClasses + deferredFunctions
     }
 
     override fun finish() {
-        // Last round, generate as much as we can, reporting errors for types that still can't be resolved.
-        for (element in deferredClasses) {
-            processInject(
-                element,
-                provider,
+        try {
+            // Last round, generate as much as we can, reporting errors for types that still can't be resolved.
+            val resolver = lastResolver ?: return
+            for (element in deferredClassNames.mapNotNull { resolver.getClassDeclarationByName(it) }) {
+                processInject(
+                    element,
+                    provider,
+                    codeGenerator,
+                    injectGenerator,
+                    skipValidation = true
+                )
+            }
+
+            for (element in deferredFunctionNames.flatMap {
+                resolver.getFunctionDeclarationsByName(
+                    it,
+                    includeTopLevel = true
+                )
+            }) {
+                processKmpComponentCreate(element, provider, kmpComponentCreateFunctionsByComponentType)
+            }
+
+            generateKmpComponentCreateFiles(
                 codeGenerator,
-                injectGenerator,
-                skipValidation = true
+                kmpComponentCreateGenerator,
+                kmpComponentCreateFunctionsByComponentType
             )
+            kmpComponentCreateFunctionsByComponentType.clear()
+        } finally {
+            lastResolver = null
+            deferredClassNames = emptyList()
+            deferredFunctionNames = mutableListOf()
         }
-        deferredClasses = mutableListOf()
-
-        for (element in deferredFunctions) {
-            processKmpComponentCreate(element, provider, kmpComponentCreateFunctionsByComponentType)
-        }
-
-        generateKmpComponentCreateFiles(
-            codeGenerator,
-            kmpComponentCreateGenerator,
-            kmpComponentCreateFunctionsByComponentType
-        )
-        kmpComponentCreateFunctionsByComponentType.clear()
-
-        deferredFunctions = mutableListOf()
```

Comment:
I'll try the changes with my breaking tests later tonight and will report back if that works.

Reply:
Seems to work great!